### PR TITLE
Uplink producer

### PR DESCRIPTION
**Pull request by athena255**

### CODE REVIEW

File: test/test_telem_info_generator.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import subprocess
```
Comment (tanishqaggarwal) [inline test/test_telem_info_generator.py]: Remove this from your change

File: src/GroundTargets/telem_info_generator.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#include "../GroundCode/TelemetryInfoGenerator.hpp"
```
Comment (tanishqaggarwal) [inline src/GroundTargets/telem_info_generator.cpp]: Remove this file

File: src/GroundCode/TelemetryInfoGenerator.hpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#ifndef FIELD_INFO_HPP_
```
Comment (tanishqaggarwal) [inline src/GroundCode/TelemetryInfoGenerator.hpp]: Remove this file

File: src/GroundCode/TelemetryInfoGenerator.cpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
+#include "TelemetryInfoGenerator.hpp"
```
Comment (tanishqaggarwal) [inline src/GroundCode/TelemetryInfoGenerator.cpp]: Remove this file

File: lib/Drivers/Piksi.cpp [hunk 5]
```diff
@@ -265,7 +265,8 @@ unsigned char Piksi::process_buffer_msg_len() {
 }
 
 unsigned char Piksi::read_all() {
-    #if defined(UNIT_TEST) || defined(DESKTOP)
+  //  #if defined(UNIT_TEST) || defined(DESKTOP) 
+    #ifdef DESKTOP
```
Comment (tanishqaggarwal) [inline lib/Drivers/Piksi.cpp]: Very confused by why these flag changes are needed. @shihaocao can you investigate? I feel like the former flags are the right ones

File: test/test_gsw_uplink_producer/test_gsw_uplink_producer.cpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    StateFieldRegistryMock registry;
+
+    std::unique_ptr<UplinkProducer> uplink_producer;
+    
+    // Create a TestFixture instance of QuakeManager with the following parameters
+    TestFixture() : registry() {
+         uplink_producer = std::make_unique<UplinkProducer>(registry);
+        // Create dummy fields
+    }
+    void check_json_registry(const char* filename)
```
Comment (tanishqaggarwal) [inline test/test_gsw_uplink_producer/test_gsw_uplink_producer.cpp]: Can you add a docstring?

File: test/test_gsw_uplink_producer/test_gsw_uplink_producer.cpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+#define TEST_ASSERT_THROW(x){try{x;TEST_ASSERT_TRUE(0);}catch(const std::exception& e){TEST_ASSERT_TRUE(1);}}
+#define TEST_ASSERT_NO_THROW(x){try{x;TEST_ASSERT_TRUE(1);}catch(const std::exception& e){TEST_ASSERT_TRUE(0);}}
+class TestFixture {
+  public:
+    StateFieldRegistryMock registry;
+
+    std::unique_ptr<UplinkProducer> uplink_producer;
+    
+    // Create a TestFixture instance of QuakeManager with the following parameters
```
Comment (tanishqaggarwal) [inline test/test_gsw_uplink_producer/test_gsw_uplink_producer.cpp]: Of "UplinkProducer"

File: test/test_gsw_uplink_producer/test_3.sbd [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (tanishqaggarwal) [inline test/test_gsw_uplink_producer/test_3.sbd]: Is this file used?

File: lib/Drivers/Piksi.cpp (hunk 5 above)
Comment (athena255) [reply lib/Drivers/Piksi.cpp]: The #if defined(UNIT_TEST) || defined (DESKTOP) breaks hardware tests.

File: src/FCCode/UplinkCommon.cpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
+#include "UplinkCommon.h"
+
+Uplink::Uplink(StateFieldRegistry& r) : registry(r)
+{
+    // calculate the maximum number of bits needed to represent the indices
+    for (index_size = 1; (registry.writable_fields.size() + 1) / (1 << index_size) > 0; ++index_size){}
```
[the remark this answers is not in the document]
Comment (athena255) [reply src/FCCode/UplinkCommon.cpp]: I will add a comment saying that the target variable is index_size.

File: src/FCCode/UplinkCommon.cpp [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+bool Uplink::_validate_packet(bitstream& bs)
+{
+    // Start validation at beginning of bs
+    bs.reset();
+    size_t packet_bytes = bs.max_len;
+    size_t field_index = 0, field_len = 0, bits_checked = 0, bits_consumed = 0;
+    // Keep a bit map to prevent updating the same field twice
+    std::vector<bool> is_field_updated(registry.writable_fields.size(), 0);
```
[the remark this answers is not in the document]
Comment (athena255) [reply src/FCCode/UplinkCommon.cpp]: This is not allocated on the heap.

File: lib/Drivers/Piksi.cpp (hunk 5 above)
Comment (shihaocao) [reply lib/Drivers/Piksi.cpp]: Which ones?

File: src/FCCode/UplinkCommon.cpp [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        // Get index from bitstream
+        bits_consumed = bs.nextN(index_size, reinterpret_cast<uint8_t*>(&field_index));
+
+        if (field_index == 0) // reached end of the packet
+            break;
+        --field_index;
+        // If we have already seen this field or if the number of bits consumed
+        // to get the next index is not index_size
+        if (is_field_updated[field_index] || bits_consumed != index_size)
```
[the remark this answers is not in the document]
Comment (athena255) [reply src/FCCode/UplinkCommon.cpp]: yes. I will move lines 34 and 35 above here.

File: src/FCCode/UplinkCommon.cpp [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        // we reached the end earlier than expected
+        if (bits_consumed != field_len) 
+            return false;
+
+        bits_checked += bits_consumed;
+    }
+
+    // Consume the padding (there should be no more than 7 bits of padding)
+    uint8_t u8 = 0;
+    bits_consumed = bs.nextN(8, &u8);
```
[the remark this answers is not in the document]
Comment (athena255) [reply src/FCCode/UplinkCommon.cpp]: It makes it easier to detect errors. The uplink packet size is variable. We do not need to send  70 bytes if we only have 55 bytes we want to send. Why send extra bytes? The uplink producer script will trim the packet to a one byte padding.

File: src/FCCode/UplinkConsumer.cpp [hunk 13]
```diff
@@ -25,103 +24,14 @@ void UplinkConsumer::execute()
     radio_mt_packet_len_fp->set(0);
```
[the remark this answers is not in the document]
Comment (athena255) [reply src/FCCode/UplinkConsumer.cpp]: I meant clear len anyway (even if the packet is invalid).  I can change this to clear len always

File: src/GroundCode/UplinkProducer.cpp [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+        auto field_p = registry.writable_fields[field_index];
+        std::cout << field_index << "\t" <<field_len;
+
+        // Dump into bit_array
+        bit_array ba(field_len);
+        bits_consumed += bs.nextN(field_len, ba);
+        std::cout << "\t" << field_p->get_bit_array().to_ulong() << " --> " <<  ba.to_ulong();
+        std::cout << "\t\t" << field_p->name() << std::endl;
+    }
+}
```
[the remark this answers is not in the document]
Comment (athena255) [reply src/GroundCode/UplinkProducer.cpp]: I used this function to debug. It prints the current packet represented by the bitstream. The changes have yet to be committed to the state fields.

File: src/GroundCode/UplinkProducer.cpp [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+void UplinkProducer::to_file(const bitstream& bs, const std::string& filename)
+{
+    bool is_valid = _validate_packet(const_cast<bitstream&>(bs));
+    // Throw exception if verification fails
+    if (!is_valid)
+       throw std::runtime_error("Uplink Producer: Packet you created is not valid");
+    // Write to file
+    std::ofstream newFile (filename, std::ios::out | std::ios::binary);
+    newFile.write(reinterpret_cast<const char*>(bs.stream), bs.max_len);
+    newFile.close();
```
[the remark this answers is not in the document]
Comment (athena255) [reply src/GroundCode/UplinkProducer.cpp]: newFile is lowercase.

File: src/GroundCode/UplinkProducer.h [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * Creates an SBD file directly from a json file
+     * @return true if successful
+     */
+    bool create_sbd_from_json(const std::string& json_file, const std::string& dst_file);
+
+    /**
+     * Return the maximum possible packet size
+     */
+    size_t get_max_possible_packet_size();
```
[the remark this answers is not in the document]
Comment (athena255) [reply src/GroundCode/UplinkProducer.h]: yes

File: src/GroundTargets/telem_info_generator.cpp (hunk 2 above)
Comment (athena255) [reply src/GroundTargets/telem_info_generator.cpp]: ok

File: test/test_gsw_uplink_producer/UplinkProducerMock.h [hunk 17]
```diff
@@ -0,0 +1,22 @@
+// #pragma once
```
[the remark this answers is not in the document]
Comment (athena255) [reply test/test_gsw_uplink_producer/UplinkProducerMock.h]: ok

File: test/test_gsw_uplink_producer/test_3.sbd (hunk 8 above)
Comment (athena255) [reply test/test_gsw_uplink_producer/test_3.sbd]: yes

File: test/test_gsw_uplink_producer/test_gsw_uplink_producer.cpp (hunk 6 above)
Comment (athena255) [reply test/test_gsw_uplink_producer/test_gsw_uplink_producer.cpp]: ok

File: test/test_telem_info_generator.py (hunk 1 above)
Comment (athena255) [reply test/test_telem_info_generator.py]: ok

File: src/FCCode/UplinkCommon.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/FCCode/UplinkCommon.cpp]: Makes sense

File: src/GroundCode/UplinkProducer.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/GroundCode/UplinkProducer.cpp]: sorry, I meant snake-case

File: test/test_gsw_uplink_producer/test_3.sbd (hunk 8 above)
Comment (tanishqaggarwal) [reply test/test_gsw_uplink_producer/test_3.sbd]: where? couldn't spot it in the test file

File: src/FCCode/UplinkCommon.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (athena255) [reply src/FCCode/UplinkCommon.cpp]: I made it static std::vector <bool> now

File: test/test_gsw_uplink_producer/test_3.sbd (hunk 8 above)
Comment (athena255) [reply test/test_gsw_uplink_producer/test_3.sbd]: Nvm, I thought you meant test_3.json. I'll delete it

**Pull request merged**